File: guides/java/kora-java-guide-resilient-app/src/main/java/ru/tinkoff/kora/guide/resilient/service/CircuitBreakerFailurePredicate.java
package ru.tinkoff.kora.guide.resilient.service;

import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.http.server.common.HttpServerResponseException;
import ru.tinkoff.kora.resilient.circuitbreaker.CircuitBreakerPredicate;

@Component
public final class CircuitBreakerFailurePredicate implements CircuitBreakerPredicate {

    @Override
    public String name() {
        return "RecordServerErrorsOnly";
    }

    @Override
    public boolean test(Throwable throwable) {
        if (throwable instanceof HttpServerResponseException exception) {
            return exception.code() >= 500;
        }
        return true;
    }
}

File: guides/kotlin/kora-kotlin-guide-resilient-app/src/main/kotlin/ru/tinkoff/kora/guide/resilient/service/CircuitBreakerFailurePredicate.kt
package ru.tinkoff.kora.guide.resilient.service

import ru.tinkoff.kora.common.Component
import ru.tinkoff.kora.http.server.common.HttpServerResponseException
import ru.tinkoff.kora.resilient.circuitbreaker.CircuitBreakerPredicate

@Component
class CircuitBreakerFailurePredicate : CircuitBreakerPredicate {
    override fun name(): String = "RecordServerErrorsOnly"

    override fun test(throwable: Throwable): Boolean {
        return throwable !is HttpServerResponseException || throwable.code() >= 500
    }
}
